Make global queue and event polling intervals configurable

aturon opened the pull request:
Closes #4651

# Motivation

Some workloads strongly benefit from changing the scheduler constants that control:

* Global queue polling
* Event polling (the driver: timer, process, and I/O events)

# Solution

Adds knobs to the runtime builder to control the number of ticks between polling the global task queue (fairness) and the event driver (I/O prioritization). The default values for these knobs are determined based on what scheduler kind the builder is started with.

Both varieties of scheduler already supported these intervals, but they were defined by private constants. This PR moves the constants into the `Core` struct, where they can be supplied by the builder configuration.

Darksonn commented on tokio/src/runtime/builder.rs:
````diff
@@ -554,6 +590,54 @@ impl Builder {
+    ///     .build();
+    /// # }
+    /// ```
+    pub fn global_queue_interval(&mut self, val: u8) -> &mut Self {
+        self.global_queue_interval = val;
+        self
+    }
````
I don't think `u8` is the right type for this. Something like `u32` would be more appropriate. We can change the internals to use an `u32` for the tick.

aturon commented:
Rebased and added [a commit](https://github.com/tokio-rs/tokio/pull/4671/commits/6e9c85036120fb4803e10174bf53177513927f32) incorporating review feedback.

Darksonn reviewed (approved):
Overall this looks pretty good.

Darksonn commented on tokio/src/runtime/builder.rs:
````diff
@@ -554,6 +590,65 @@ impl Builder {
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use tokio::runtime;
+    ///
+    /// # pub fn main() {
````
This empty line shows up on the documentation. We should either remove the empty line or unhide the import.

(The same applies to `event_interval`, and apparently also the existing `thread_keep_alive`)

Darksonn replied:
Actually, it seems like it happens on quite a few methods on the `Builder`. It might be better to leave for another PR.

aturon replied:
Fixed in https://github.com/tokio-rs/tokio/pull/4671/commits/819998cffcf0f35d2b3affa323d06a85745f9182

aturon commented:
I've pushed up some further changes:

* Removed all extraneous blank lines in doc comments.
* Cleaned up the way the defaults are defined.
* Included default values in the API docs.

aturon commented on tokio/src/runtime/builder.rs:
```diff
@@ -554,6 +589,63 @@ impl Builder {
+    /// getting started on new work, especially if tasks frequently yield rather than complete
+    /// or await on further I/O. Conversely, a higher value prioritizes existing work, and
+    /// is a good choice when most tasks quickly complete polling.
+    ///
```
Fixed in https://github.com/tokio-rs/tokio/pull/4671/commits/793a09e2c627dee0301f0670f0bdf915fb24d6fd

hawkw replied:
thanks!

aturon commented on tokio/src/runtime/builder.rs:
```diff
@@ -554,6 +589,63 @@ impl Builder {
+    /// so if tasks generally complete their polling quickly, a higher event interval
+    /// will minimize that overhead while still keeping the scheduler responsive to
+    /// events.
+    ///
```
Fixed in https://github.com/tokio-rs/tokio/pull/4671/commits/793a09e2c627dee0301f0670f0bdf915fb24d6fd

hawkw reviewed (approved):
docs changes look good to me, thanks!

Darksonn commented:
Thanks!

aturon commented:
Just to check in on this -- what is the typical release cadence? I'd like to pull this in as an official release rather than a github patch if possible, but not sure how soon a release including it might happen.

Darksonn commented:
We don't have a fixed release cadence, but looking at the commit history, it does seem like it is about time to make another release.
